### ISSUE TITLE
Security hardening: replace unsafe eval-based scheduler expression parsing with a restricted arithmetic AST evaluator (CWE-95)

### ISSUE THREAD

**Issue by abhayrajjais01** [weecology/DeepForest#1305]
### Summary

`configure_optimizers()` in [`src/deepforest/main.py`](https://github.com/weecology/DeepForest/blob/main/src/deepforest/main.py) currently evaluates the `lr_lambda` scheduler parameter using bare `eval()`:

```python
def lr_lambda(epoch):
    return eval(params.lr_lambda)
```

Because `params.lr_lambda` is a user-configurable string (via YAML config or Hydra/OmegaConf CLI overrides), this allows **arbitrary Python code execution** during optimizer setup for the `lambdaLR` and `multiplicativeLR` scheduler types.

This maps to [CWE-95: Improper Neutralization of Directives in Dynamically Evaluated Code](https://cwe.mitre.org/data/definitions/95.html).

---

### Exploit preconditions

This is **not** a remotely exploitable vulnerability in typical usage. To trigger it, an attacker needs:

1. **Control over the config file or CLI overrides** — e.g., a malicious YAML config, environment variable injection, or Hydra override
2. **The scheduler type set to `lambdaLR` or `multiplicativeLR`** — other scheduler types do not use `eval()`

**Impact is strongest in:**
- Shared Jupyter notebook hubs (JupyterHub, Google Colab) where configs may come from untrusted sources
- CI/CD pipelines where config files are pulled from external repositories
- Supply-chain scenarios where a dependency or dataset bundle includes a config file

In single-user, self-managed environments the risk is lower, but this remains a defense-in-depth concern for the broader scientific software ecosystem.

---

### Minimal reproduction

```python
from deepforest import main
import torch

config_args = {
    "train": {
        "lr": 0.01,
        "scheduler": {
            "type": "lambdaLR",
            "params": {
                "lr_lambda": "__import__('os').system('echo ARBITRARY_CODE_EXECUTED')",
            },
        },
        "csv_file": "dummy.csv",
        "root_dir": ".",
        "fast_dev_run": False,
    },
    "validation": {"csv_file": None, "root_dir": "."},
}

m = main.deepforest(model=torch.nn.Linear(1, 1), config_args=config_args)
m.configure_optimizers()  # ← executes os.system(...)
```

On current `main`, this will print `ARBITRARY_CODE_EXECUTED` to the console (or execute whatever payload is substituted).

---

### Proposed fix

Replace `eval()` with a **restricted AST-based evaluator** that only permits:

| Allowed | Examples |
|---|---|
| Numeric constants | `0.95`, `10`, `1e-3` |
| The variable `epoch` | `epoch` |
| Arithmetic operators | `+`, `-`, `*`, `/`, `//`, `%`, `**` |
| Unary `+` / `-` | `-0.1`, `+epoch` |
| Parentheses | `(epoch + 1)` |

Anything else (function calls, attribute access, imports, subscripts, etc.) raises a `ValueError("Unsafe lr_lambda expression: ...")`.

This preserves all **valid** scheduler expressions (e.g., `0.95 ** epoch`, `1 - epoch * 0.01`) while blocking executable payloads.

A **regression test** should be included that injects a malicious string and asserts `ValueError` is raised.

---

### Related issues

- **#1258** — [Scheduler inconsistencies in config](https://github.com/weecology/DeepForest/issues/1258): This fix touches the same `configure_optimizers()` code path. The `eval()` replacement can be coordinated with the broader scheduler config cleanup discussed there.

---

I have used ChatGpt for explaination of this issue

###

I'm happy to open a PR with:
- The `_safe_eval_lr_lambda()` static method implementation
- Updated `configure_optimizers()` call site
- A targeted regression test (`test_configure_optimizers_rejects_unsafe_lr_lambda`)

Attaching a pr below

Let me know if you like me to go with or if you want me to make any changes
